<a href="https://colab.research.google.com/github/PauloRhyanK/ProjetoeAnaliseAlgoritimos/blob/main/Extra_2_An%C3%A1lise_de_Complexidade_Exponencial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


<html>
  <body>
    <header></header>
        <CENTER>
          <img src="https://uvv.br/wp-content/themes/core/dist/images/Logo@2x.png" alt="UVV-LOGO" style = width="100px"; height="100px">
        </CENTER>
        <CENTER>
          <b>
            <font size="5">PROJETO E ANÁLISE DE ALGORITMOS
          </b>
          </CENTER>
        <CENTER><b>ANÁLISE DE COMPLEXIDADE EXPONENCIAL</b></CENTER><br/>
        </font>

COLAB: [Link]()

# BIBLIOTECAS

In [15]:
import numpy as np
import time
import matplotlib.pyplot as plt
import graphviz

# FUNÇÕES

# EXERCÍCIOS

Fazer todas as fórmulas matemática usando a Linguagem de Marcação: $L^aTe_x$

# 1) TÍTULO: Análise de Complexidade Exponencial

A análise de complexidade exponencial refere-se à avaliação de algoritmos cujo tempo de execução (ou espaço) cresce exponencialmente com o tamanho da entrada. Em termos práticos, isso significa que, conforme o tamanho dos dados de entrada aumenta, os recursos necessários para resolver o problema aumentam de forma extremamente rápida – tipicamente, da ordem de O(2ⁿ) ou similar.

### Conceitos-Chave

- **Tempo de Execução:**  
  Em algoritmos exponenciais, dobrar o tamanho da entrada pode, por exemplo, elevar o tempo de execução a uma potência, tornando impraticável resolver instâncias maiores.

- **Exemplo Clássico:**  
  Um exemplo é o algoritmo de força bruta para o problema do caixeiro viajante, onde a quantidade de caminhos a serem avaliados cresce exponencialmente conforme o número de cidades aumenta.

- **Comparação com Outras Complexidades:**  
  - **Polinomial (por exemplo, O(n²), O(n³)):** O tempo cresce de forma mais “suave” e escalável.
  - **Exponencial:** O crescimento é muito mais acentuado, levando a tempos de execução inviáveis mesmo para entradas moderadamente grandes.

### Importância na Ciência da Computação

A análise de algoritmos exponenciais é fundamental para:
- **Identificar Limitações:** Compreender os limites práticos de um algoritmo e identificar quando uma abordagem pode não ser escalável.
- **Desenvolver Algoritmos mais Eficientes:** Incentivar a busca por métodos aproximados ou heurísticos que possam oferecer soluções aceitáveis sem explorar todo o espaço exponencial de possibilidades.

### Estratégias para Lidar com Complexidade Exponencial

1. **Heurísticas e Algoritmos Aproximados:**  
   Em muitos problemas onde a solução ótima é difícil de ser encontrada em tempo razoável, algoritmos aproximados podem oferecer soluções “boas o suficiente” de forma mais rápida.

2. **Programação Dinâmica:**  
   Alguns problemas inicialmente exponenciais podem ser otimizados para uma complexidade polinomial através da reutilização de subsoluções já calculadas.

3. **Divisão e Conquista:**  
   Em certos casos, a técnica de dividir o problema em subproblemas menores pode reduzir a complexidade prática, mesmo que a solução completa ainda tenha um componente exponencial.

### Conclusão

A análise de complexidade exponencial é essencial para entender e prever o comportamento de algoritmos em termos de tempo e recursos necessários, especialmente para problemas que podem ser inviáveis em larga escala. Reconhecer um algoritmo exponencial permite aos desenvolvedores e pesquisadores buscar estratégias alternativas para melhorar o desempenho e a viabilidade prática dos métodos implementados.

# 2) O algoritmo exponencial do Simplex em Pseudocódigo e Python (Executar o algoritmo para No + 2 instâncias)



```
 Algoritmo SIMPLEX(c, A, b)
Entrada:
    c: vetor de custos (dimensão n)
    A: matriz de restrições (m x n)
    b: vetor dos termos independentes (dimensão m)
    
1. Converter o problema para a forma padrão:
   a. Adicionar m variáveis de folga para transformar cada inequação em igualdade.
   b. Construir a tabela simplex inicial:
         - Linhas 1 a m: coeficientes de A, matriz identidade (para folgas) e coluna b.
         - Última linha: -c (coeficientes da função objetivo) e zeros para as folgas e o lado direito.

2. Inicializar o conjunto de variáveis básicas com os índices das variáveis de folga.

3. Enquanto existir alguma coluna na linha de custo (linha da função objetivo) com valor negativo:
   a. Selecionar a coluna j* com o menor valor (mais negativo) na linha de custo.
   b. Para cada linha i onde o elemento na coluna j* é positivo, calcular a razão:
          razão[i] = (elemento na coluna RHS da linha i) / (A[i][j*]).
   c. Se nenhuma razão puder ser calculada (ou seja, nenhum elemento positivo em j*), então:
          → Retorne "Problema Ilimitado".
   d. Selecione a linha i* com o menor valor da razão (teste da razão).
   e. Realize o pivô na posição (i*, j*):
          - Divida a linha i* pelo elemento pivô.
          - Para cada outra linha, subtraia um múltiplo adequado da linha pivô para zerar o elemento na coluna j*.
          - Atualize o conjunto de variáveis básicas, substituindo a variável que estava na linha i* pela variável j*.

4. Ao final, quando não houver mais custo negativo, extraia a solução:
   - Para cada variável de decisão (não-folga), se ela for básica, seu valor é o correspondente na coluna RHS.
   - O valor ótimo é dado na última coluna da linha de custo.

5. Retorne a solução ótima e o valor da função objetivo.
```



In [23]:
import numpy as np

def simplex(c, A, b):
    """
    Implementação básica do método Simplex para problemas de maximização.
    Entradas:
      c : vetor de custos (dimensão n)
      A : matriz de restrições (m x n)
      b : vetor do lado direito (dimensão m)
    Retorna:
      x : vetor solução (dimensão n)
      optimal_value : valor ótimo da função objetivo
    """
    m, n = A.shape

    # Construir a tabela simplex: dimensões (m+1) x (n + m + 1)
    # Colunas: variáveis de decisão (n), variáveis de folga (m) e coluna RHS
    tableau = np.zeros((m+1, n+m+1))

    # Preencher as linhas de restrição: A, identidade para folgas e coluna b
    tableau[:m, :n] = A
    tableau[:m, n:n+m] = np.eye(m)
    tableau[:m, -1] = b

    # Linha da função objetivo: usamos -c (para maximização)
    tableau[m, :n] = -c

    # Conjunto inicial de variáveis básicas (índices das variáveis de folga)
    basis = list(range(n, n+m))

    # Loop principal do Simplex
    while True:
        # Escolher variável de entrada: coluna com menor valor na linha objetivo
        col = np.argmin(tableau[m, :n+m])
        if tableau[m, col] >= 0:
            # Não há mais melhoria: solução ótima encontrada
            break

        # Teste da razão: determinar qual variável sairá da base
        ratios = []
        for i in range(m):
            if tableau[i, col] > 0:
                ratios.append(tableau[i, -1] / tableau[i, col])
            else:
                ratios.append(np.inf)
        row = np.argmin(ratios)
        if ratios[row] == np.inf:
            raise Exception("Problema Ilimitado")

        # Operação de pivô na posição (row, col)
        pivot = tableau[row, col]
        tableau[row, :] = tableau[row, :] / pivot
        for i in range(m+1):
            if i != row:
                tableau[i, :] -= tableau[i, col] * tableau[row, :]

        # Atualizar a base: a variável da coluna 'col' passa a ser básica na linha 'row'
        basis[row] = col

    # Extrair a solução: para as variáveis de decisão (colunas 0 a n-1)
    x = np.zeros(n)
    for i in range(m):
        if basis[i] < n:
            x[basis[i]] = tableau[i, -1]
    optimal_value = tableau[m, -1]
    return x, optimal_value

def gerar_instance_klee_minty(n):
    """
    Gera uma instância do problema de Klee-Minty com n variáveis.
    Retorna:
      c : vetor de custos (dimensão n)
      A : matriz de restrições (n x n)
      b : vetor do lado direito (dimensão n)
    O problema é:
      Maximizar ∑ (j=0 a n-1) 10^(n-1-j) * x_j
      sujeito a, para i=0 a n-1:
          ∑ (j=0 a i) 10^(i-j) * x_j <= 10^(i)
          x_j >= 0
    """
    A = np.zeros((n, n))
    b = np.zeros(n)
    for i in range(n):
        b[i] = 10**(i)  # b[i] = 10^i
        for j in range(i+1):
            A[i, j] = 10**(i - j)
    # Vetor de custos: c[j] = 10^(n-1-j)
    c = np.array([10**(n-1-j) for j in range(n)])
    return c, A, b


# Definir n0 e executar para n0 e n0+2 variáveis
n0 = 2
for n in [n0, n0+2]:
    print(f"\nExecutando Simplex para n = {n}")
    c, A, b = gerar_instance_klee_minty(n)
    try:
        x, optimal_value = simplex(c, A, b)
        print("Solução ótima encontrada:")
        print("x =", x)
        print("Valor ótimo =", optimal_value)
    except Exception as e:
        print("Erro:", e)




Executando Simplex para n = 2
Solução ótima encontrada:
x = [1. 0.]
Valor ótimo = 10.0

Executando Simplex para n = 4
Solução ótima encontrada:
x = [1. 0. 0. 0.]
Valor ótimo = 1000.0


In [21]:
def simplex(c, A, b):
    """
    Implementação básica do método Simplex para problemas de maximização.
    """
    m, n = A.shape

    # Construir a tabela simplex: dimensões (m+1) x (n + m + 1)
    tableau = np.zeros((m+1, n+m+1))

    # Preencher as linhas de restrição: A, identidade para folgas e coluna b
    tableau[:m, :n] = A
    tableau[:m, n:n+m] = np.eye(m)
    tableau[:m, -1] = b

    # Linha da função objetivo: usamos -c (para maximização)
    tableau[m, :n] = -c

    # Conjunto inicial de variáveis básicas (índices das variáveis de folga)
    basis = list(range(n, n+m))

    # Loop principal do Simplex
    while True:
        col = np.argmin(tableau[m, :n+m])
        if tableau[m, col] >= 0:
            break

        # Teste da razão: determinar qual variável sairá da base
        ratios = []
        for i in range(m):
            if tableau[i, col] > 0:
                ratios.append(tableau[i, -1] / tableau[i, col])
            else:
                ratios.append(np.inf)
        row = np.argmin(ratios)
        if ratios[row] == np.inf:
            raise Exception("Problema Ilimitado")

        # Operação de pivô na posição (row, col)
        pivot = tableau[row, col]
        tableau[row, :] /= pivot
        for i in range(m+1):
            if i != row:
                tableau[i, :] -= tableau[i, col] * tableau[row, :]

        # Atualizar a base
        basis[row] = col

    # Extrair a solução
    x = np.zeros(n)
    for i in range(m):
        if basis[i] < n:
            x[basis[i]] = tableau[i, -1]
    optimal_value = tableau[m, -1]
    return x, optimal_value



In [22]:
# Definição da entrada
A = np.array([[-1, 5, -2],
              [ 2, -3,  7],
              [-7, 9, -6]], dtype=float)
b = np.array([10, -2, 8], dtype=float)

# Função objetivo fictícia (podemos maximizar a soma das variáveis)
c = np.array([1, 1, 1], dtype=float)

# Resolver o problema
x, optimal_value = simplex(c, A, b)

# Exibir os resultados
print("Solução do sistema via Simplex:")
print(f"x = {x[0]:.3f}")
print(f"y = {x[1]:.3f}")
print(f"z = {x[2]:.3f}")
print(f"Valor ótimo = {optimal_value:.3f}")

Solução do sistema via Simplex:
x = 2.857
y = 2.571
z = 0.000
Valor ótimo = 5.429




## Explicação

- **Geração da instância (função `gerar_instance_klee_minty`):**  
  Essa função constrói uma matriz \( A \) triangular inferior em que os coeficientes são potências de 10, define o vetor \( b \) com termos \( 10^i \) e o vetor de custos \( c \) de forma que a função objetivo seja, por exemplo,  
  \[
  \text{Maximizar } 10^{n-1} x_1 + 10^{n-2} x_2 + \cdots + 1\cdot x_n.
  \]
  Essa formulação é inspirada no exemplo de Klee-Minty, conhecido por forçar o algoritmo Simplex a explorar um caminho exponencial em casos piores.

- **Implementação do Simplex:**  
  A função `simplex` monta uma tabela (tableau) que inclui as variáveis de decisão e as de folga. Em cada iteração, ela seleciona a variável que melhor melhora o valor da função objetivo (usando o teste do menor coeficiente na linha objetivo) e determina qual variável sairá da base pelo teste da razão. Em seguida, realiza o pivô e repete o processo até que não existam mais coeficientes negativos na linha de custo.

- **Execução para duas instâncias:**  
  No `main()`, o código executa o Simplex para uma instância com \( n_0 \) variáveis (neste exemplo, 2) e outra com \( n_0+2 \) (ou seja, 4 variáveis). Essa abordagem pode ajudar a visualizar como o algoritmo se comporta conforme o tamanho do problema aumenta – lembrando que, em casos patológicos, o número de iterações pode crescer de forma exponencial.

---

Este exemplo serve para demonstrar tanto a ideia do algoritmo Simplex quanto como certas instâncias (como a de Klee-Minty) podem levar a um comportamento exponencial em termos de iterações, apesar de o algoritmo ser geralmente eficiente na prática.

# 3) Solicitar a Análise de Complexidade e Notação Assintótica: Pior, Melhor e Caso Médio.

#### **Análise de Complexidade do Algoritmo Simplex**
O algoritmo Simplex é um método iterativo para resolver problemas de Programação Linear (PL). Seu tempo de execução depende da quantidade de variáveis, restrições e da escolha dos pivôs ao longo das iterações. Existem três cenários principais para sua análise de complexidade:

- **Melhor Caso**: Quando a solução ótima é encontrada rapidamente, o número de iterações pode ser **linear** no número de restrições.
- **Pior Caso**: Para instâncias especialmente construídas (como o exemplo de Klee-Minty), o algoritmo pode percorrer um número **exponencial** de vértices.
- **Caso Médio**: Na prática, o Simplex se comporta de maneira **pseudopolinomial**, resolvendo muitos problemas rapidamente.

A seguir, analisamos cada caso detalhadamente.

---

#### **1. Pior Caso: O(2ⁿ)**
O pior caso ocorre em instâncias patológicas, como as construídas por **Klee e Minty**, onde o algoritmo Simplex percorre **todos os vértices do politopo** (caminho exponencial).  

- O número de iterações no pior caso pode chegar a **O(2ⁿ)**, onde **n** é o número de variáveis.
- Isso ocorre porque o Simplex segue um caminho degenerado ao longo dos vértices de um politopo altamente distorcido.
- Este comportamento foi demonstrado por Klee-Minty com um politopo que força o Simplex a explorar **todos os 2ⁿ vértices** antes de encontrar a solução ótima.

📌 **Exemplo prático**:  
Se tivermos **n = 10** variáveis, o número de iterações pode chegar a **2¹⁰ = 1024**, tornando-se inviável para valores maiores de **n**.

---

####  **2. Melhor Caso: O(m)**
O melhor caso ocorre quando:
- A escolha da variável pivô é feita de maneira ótima.
- A solução é encontrada em poucas iterações, independentemente do número de variáveis.

Neste caso:
- O algoritmo encontra a solução em **no máximo m** iterações, onde **m** é o número de restrições.
- O tempo de cada iteração depende da **escolha do pivô**, que geralmente pode ser feito em **O(nm)** operações.
- Portanto, no melhor caso, **a complexidade total é O(mn)**.

📌 **Exemplo prático**:  
Se tivermos um problema com **100 variáveis e 50 restrições**, e a solução for encontrada rapidamente, o número de iterações será **≤ 50**, resultando em tempo **O(50 × 100) = O(5000)**.

---

#### **3. Caso Médio: O(n³) ~ O(n⁶)**
O caso médio é um pouco mais difícil de definir, pois depende da distribuição estatística dos problemas de PL. Contudo, estudos práticos e experimentais indicam que:

- Para **m restrições** e **n variáveis**, o número médio de iterações é **O(n) ou O(n²)**.
- O custo de cada iteração envolve:
  - Encontrar a variável pivô: **O(mn)**.
  - Atualizar a tabela simplex: **O(nm)**.
  - Resolver o sistema linear associado (eliminação de Gauss): **O(n²)**.

Assim, a complexidade do Simplex no caso médio geralmente é **entre O(n³) e O(n⁶)**.

📌 **Exemplo prático**:  
Para **n = 100**, o número de iterações esperado pode ser **100 ou 10.000**, levando a complexidade **O(100³) = O(1.000.000)** no pior cenário do caso médio.

---

#### **Resumo da Notação Assintótica**
| Caso         | Número de Iterações | Custo por Iteração | Complexidade Total |
|-------------|--------------------|-------------------|--------------------|
| Melhor Caso | O(n)               | O(nm)            | O(mn)             |
| Caso Médio  | O(n) ~ O(n²)        | O(nm)            | O(n³) ~ O(n⁶)     |
| Pior Caso   | O(2ⁿ)               | O(nm)            | O(2ⁿ)             |

#### **Conclusões**
- O Simplex **funciona muito bem na prática** e é amplamente utilizado em problemas reais.
- Em casos normais, ele se comporta **quase polinomialmente (O(n³) ~ O(n⁶))**.
- No **pior caso (O(2ⁿ))**, ele pode ser impraticável, tornando necessário o uso de algoritmos alternativos como **o método do ponto interior (que tem complexidade polinomial O(n³))**.

🚀 **Conclusão final:**  
Apesar do pior caso exponencial, o Simplex continua sendo um dos algoritmos mais usados devido ao seu desempenho eficiente na maioria dos problemas práticos.

# 4) Para o Pior Caso (Notação Big O): A Função de Complexidade exata deste algoritmo.

Para determinar a **função de complexidade exata** do algoritmo Simplex no pior caso, precisamos considerar quantas operações são realizadas por iteração e quantas iterações podem ocorrer no pior cenário.

---

#### **1. Fórmula de Complexidade no Pior Caso**
No pior caso, o Simplex pode visitar **todos os vértices do politopo viável**, o que resulta em um comportamento **exponencial**. Isso foi demonstrado por Klee e Minty (1972), que construíram um exemplo onde o Simplex percorre **2ⁿ - 1** vértices antes de encontrar a solução ótima.

A complexidade exata pode ser expressa como:

\[
T(n) = (2^n - 1) \cdot O(n^2)
\]

onde:
- \( 2^n - 1 \) representa o **número máximo de vértices visitados**.
- \( O(n^2) \) corresponde ao **custo por iteração**, que vem da atualização da tabela simplex (pivotação, escolha da variável de entrada/saída e ajustes na matriz).

Portanto, a função exata para o tempo de execução do Simplex no pior caso é:

\[
T(n) = O(2^n \cdot n^2)
\]

---

#### **2. Justificativa**
Cada iteração do Simplex envolve:
1. **Escolha da variável pivô** → **O(n)**
2. **Teste da razão (identificar a variável que sai da base)** → **O(n)**
3. **Pivotação (atualização da tabela, eliminação de Gauss para resolver sistema linear)** → **O(n^2)**

Assim, o tempo total por iteração é aproximadamente **O(n^2)**.

No **pior caso**, o algoritmo pode visitar **2ⁿ - 1** vértices antes de parar, resultando em **(2ⁿ - 1) iterações**.

Multiplicando pelo custo por iteração, temos:

\[
T(n) = (2^n - 1) \cdot O(n^2)
\]

O termo \(-1\) é irrelevante para a notação assintótica, então podemos simplificar para:

\[
T(n) = O(2^n \cdot n^2)
\]

---

#### **3. Interpretação**
- O Simplex **cresce exponencialmente** no pior caso.
- Isso significa que **mesmo para pequenos valores de \( n \)**, o tempo de execução pode ser impraticável.
- **Por exemplo**, para \( n = 20 \), temos aproximadamente **\( O(2^{20} \cdot 400) \approx O(419.4 \text{ milhões}) \)** operações.
- **Alternativa:** O **método do ponto interior** resolve problemas de Programação Linear com complexidade **polinomial O(n³)** e evita essa explosão exponencial.

🚀 **Conclusão**: No pior caso, a complexidade exata do Simplex é **O(2ⁿ \cdot n²)**, tornando-o impraticável para grandes valores de **n**.

# 5) Resolvendo a conta

A seguir, temos uma solução modularizada em Python para resolver o sistema

\[
\begin{cases}
-x + 5y - 2z = 10,\\[1mm]
2x - 3y + 7z = -2,\\[1mm]
-7x + 9y - 6z = 8.
\end{cases}
\]

No exemplo, definimos funções separadas para:
1. Resolver o sistema linear usando o NumPy.
2. Calcular os valores das expressões (funções) de cada equação, para conferir que o sistema foi resolvido corretamente.

---

### Código Python



---


In [17]:
def solve_linear_system():
    """
    Resolve o sistema linear:
      -x + 5y - 2z = 10
      2x - 3y + 7z = -2
      -7x + 9y - 6z = 8
    Retorna:
      Uma tupla (x, y, z) com os valores encontrados.
    """
    # Matriz dos coeficientes (A) e vetor dos termos independentes (b)
    A = np.array([[-1, 5, -2],
                  [ 2, -3,  7],
                  [-7, 9, -6]], dtype=float)
    b = np.array([10, -2, 8], dtype=float)
    # Resolve o sistema A * X = b
    solution = np.linalg.solve(A, b)
    return solution

def compute_functions(x, y, z):
    """
    Calcula os valores das funções definidas pelas equações do sistema:
      f1 = -x + 5y - 2z  (deve resultar em 10)
      f2 = 2x - 3y + 7z   (deve resultar em -2)
      f3 = -7x + 9y - 6z  (deve resultar em 8)
    Retorna:
      Uma tupla (f1, f2, f3) com os valores calculados.
    """
    f1 = -x + 5*y - 2*z
    f2 =  2*x - 3*y + 7*z
    f3 = -7*x + 9*y - 6*z
    return f1, f2, f3


# Resolve o sistema e obtém os valores de x, y e z
x, y, z = solve_linear_system()

# Calcula os valores das funções para conferência
f1, f2, f3 = compute_functions(x, y, z)

# Exibe os resultados
print("Solução do sistema:")
print(f"  x = {x:.3f}")
print(f"  y = {y:.3f}")
print(f"  z = {z:.3f}")

print("\nVerificação (substituindo na equação):")
print(f"  -x + 5y - 2z = {f1:.3f}  (deve ser 10)")
print(f"   2x - 3y + 7z = {f2:.3f}  (deve ser -2)")
print(f"  -7x + 9y - 6z = {f3:.3f}  (deve ser 8)")


Solução do sistema:
  x = 1.806
  y = 2.463
  z = 0.254

Verificação (substituindo na equação):
  -x + 5y - 2z = 10.000  (deve ser 10)
   2x - 3y + 7z = -2.000  (deve ser -2)
  -7x + 9y - 6z = 8.000  (deve ser 8)



### Resultados Obtidos

Ao executar o código, obtemos:

- **x** = 121/67 ≈ 1.806  
- **y** = 165/67 ≈ 2.463  
- **z** = 17/67 ≈ 0.254

E, ao substituir esses valores nas equações, temos:

- \(-x + 5y - 2z \approx 10\)
- \(2x - 3y + 7z \approx -2\)
- \(-7x + 9y - 6z \approx 8\)

Assim, os resultados finais (x, y, z, f) podem ser interpretados como:

- **x** = 1.806  
- **y** = 2.463  
- **z** = 0.254  
- **f** = 10 (valor da função da 1ª equação, por exemplo)

Esses valores confirmam que o sistema foi resolvido corretamente.